pkg/controller/deployment_machineset_util.go

```diff
@@ -109,6 +113,25 @@ func calculateMachineSetStatus(is *v1alpha1.MachineSet, filteredMachines []*v1al
 				readyReplicasCount++
 			}
 		}
+		if machine.Status.LastOperation.State == v1alpha1.MachineStateFailed {
+			machineSummary.Name = machine.Name
+			machineSummary.ProviderID = machine.Spec.ProviderID
+			machineSummary.LastOperation = machine.Status.LastOperation
+			//ownerRef populated here, so that deployment controller doesn't have to add it seperately
+			if controller := metav1.GetControllerOf(machine); controller != nil {
+				machineSummary.OwnerRef = controller.Name
+			}
+			failedMachines = append(failedMachines, machineSummary)
+		}
+	}
+
+	// Update the FailedMachines field only if we see new failures
+	// Clear FailedMachines if ready replicas equals total replicas,
+	// which means the machineset doesn't have any machine objects which are in any failed state
+	if len(failedMachines) > 0 {
+		newStatus.FailedMachines = &failedMachines
+	} else if int32(readyReplicasCount) == is.Status.Replicas {
```

Comment:
Why relying on the readyReplicasCount ?
Is there a possibility of having enough ready machine-objects but few machines stuck with an error in the control-plane?